sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -363,6 +364,70 @@ trait Row extends Serializable {
     false
   }
 
+  protected def canEqual(other: Any) =
+    other.isInstanceOf[Row] && !other.isInstanceOf[InternalRow]
```

Comment:
So still leave the `equals` / `hashCode` implemented in `InternalRow` seems more reasonable?